extractor/src/main/java/org/schabi/newpipe/extractor/uih/UIHFactory.java

```diff
@@ -23,52 +23,46 @@
  * along with NewPipe.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-public abstract class UrlIdHandler {
+public abstract class UIHFactory {
```

Comment:
Ideally, I would keep this as an interface and use default method for `fromUrl` and others which can be used by the implementor. Always go for composition over inheritence when the code is still simple to refactor ;)

This also means we might want to do to same to `UIHandler` and make it implement a `Linkable` that only has the `originalUrl` and `url` and `Identifiable` that contains the `id`. This way, `ListUIHandler` no longer have to extend `UIHandler`, and instead implements `Linkable`, `Identifiable` and `Filterable`.

This gives us more flexibility later down the road and also helps addressing potential misuses where the `id` in `ListUIHandler` is used as a holder for the search query string in `SearchQIHandler`, which would be decoupled from `ListUIHandler` and instead implements `Linkable` and `Filterable` and the search query becomes a field.

I understand this is probably a lot of work, so I'm fine with keeping the inheritance structure for now. However, I recommend having it done right now as there will likely be a time when more complex url uses arise when simple inheritance will not be enough.

Reply:
> I understand this is probably a lot of work, so I'm fine with keeping the inheritance structure for now. However, I recommend having it done right now as there will likely be a time when more complex url uses arise when simple inheritance will not be enough.

Yea your ideas are quite good, but the refactor took me almost two months and now I'm already done. I don't want to throw all over and start again.

However we should consider something like this, because once we pull off multisservice support and everyone starts writing supporting services there is probably no turning back (or it will be hart).